docs/rules/func-names.md

```diff
@@ -45,6 +46,34 @@ Foo.prototype.bar = function bar() {};
 }())
 ```
 
+### as-needed
+
+In ES6, aside from explicit names, function names may be come from the surrounding context of the function. For example, a function assigned to a variable will automatically use the name of the variable as its name if it have no name of its own.
+
+Examples of **incorrect** code for this rule with the default `"as-needed"` option:
+
+```js
+/*eslint func-names: ["error", "as-needed"]*/
+
+Foo.prototype.bar = function() {};
```

Comment:
I'm going to link [this discussion](https://github.com/nodejs/node/pull/9252#pullrequestreview-5450499), because it seems to be relevant here. Basically, naming the function in a case like this might actually result in *less* information in the stack trace.

That said, the function doesn't have an inferred name, so I suppose it's correct to still enforce this.